pkg/util/deeptest/statictesthelper.go

```diff
@@ -0,0 +1,256 @@
+					h.assertEveryFieldNotEqual(t, valA.Index(i), valB.Index(i), path+fmt.Sprintf("[%d]", i))
+				}
+			}
+		}
+	case reflect.Array:
+		if valA.Len() == valB.Len() {
+			for i := 0; i < valA.Len(); i++ {
+				h.assertEveryFieldNotEqual(t, valA.Index(i), valB.Index(i), path+fmt.Sprintf("[%d]", i))
+			}
+		}
+	case reflect.Bool:
+		require.NotEqual(t, valA.Bool(), valB.Bool(), path+" should not be the same")
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		require.NotEqual(t, valA.Int(), valB.Int(), path+" should not be the same")
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		require.NotEqual(t, valA.Uint(), valB.Uint(), path+" should not be the same")
+	case reflect.Float32, reflect.Float64:
+		require.NotEqual(t, valA.Float(), valB.Float(), path+" should not be the same")
+	case reflect.String:
+		require.NotEqual(t, valA.String(), valB.String(), path+" should not be the same")
+	case reflect.Map:
+		require.NotEqual(t, valA.Pointer(), valB.Pointer(), path+" should not be the same")
+
+		if !h.shouldComparePointer(path) {
+			if valA.Len() == valB.Len() {
+				for _, key := range valA.MapKeys() {
+					h.assertDeepClonedEqual(t, valA.MapIndex(key), valB.MapIndex(key), path+fmt.Sprintf("[%v]", key))
+				}
+			}
+		}
+	case reflect.Interface:
+		if valA.IsNil() || valB.IsNil() {
+			require.False(t, valA.IsNil() && valB.IsNil(), path+" should not be nil at the same time")
+			return
+		}
+		h.assertEveryFieldNotEqual(t, valA.Elem(), valB.Elem(), path)
+	case reflect.Func:
+		if !h.shouldComparePointer(path) && !h.shouldIgnore(path) {
+			require.Fail(t, "a function should be compared by pointer or ignored, because there's no way to compare the content of a function", path)
+		}
+
+		require.NotEqual(t, valA.Pointer(), valB.Pointer(), path+" should be different")
+		return
+	default:
+		require.Fail(t, "unsupported type", path)
+	}
+}
+
+func (h *staticTestHelper) assertDeepClonedEqual(t *testing.T, valA, valB reflect.Value, path string) {
+	if h.shouldIgnore(path) {
+		return
+	}
+
+	// This function assumes that `a` and `b` are the same type
+	switch valA.Type().Kind() {
+	case reflect.Struct:
+		for i := 0; i < valA.NumField(); i++ {
+			h.assertDeepClonedEqual(t, valA.Field(i), valB.Field(i), path+"."+valA.Type().Field(i).Name)
+		}
+	case reflect.Ptr:
+		// both of them are not nil
+		require.NotEqual(t, 0, valA.Pointer(), path+" should not be nil")
+		require.NotEqual(t, 0, valB.Pointer(), path+" should not be nil")
+
+		if h.shouldComparePointer(path) {
+			require.Equal(t, valA.Pointer(), valB.Pointer(), path+" should be the same")
+		} else {
+			require.NotEqual(t, valA.Pointer(), valB.Pointer(), path+" should be different")
+			h.assertDeepClonedEqual(t, valA.Elem(), valB.Elem(), path)
+		}
+	case reflect.Slice:
+		if h.shouldComparePointer(path) {
+			require.Equal(t, valA.Pointer(), valB.Pointer(), path+" should be the same")
```

Comment:
Do we need also to compare the lenght of these two slice?

Reply:
Yes.